### PR TITLE
Make hoppers respect inventory max stack size

### LINKED ISSUE #2662: Hoppers don't respect MaxStackSize value of container
### What behaviour is expected:
```java
// BlockDropper is instance of org.bukkit.block.Dropper
BlockDropper.getInventory().setMaxStackSize(1);
```
Items in dropper don't stack when being transferred from the hopper.

### What behaviour is observed:
Items stack up in the first slot of the dropper. However, manually placing the items works as intended.

### Steps/models to reproduce:
1. Create plugin that sets Dropper's MaxStackSize to 1
```java
// BlockDropper is instance of org.bukkit.block.Dropper
BlockDropper.getInventory().setMaxStackSize(1);
```
2. Set dropper's MaxStackSize to 1.
3. Place hopper above the dropper.
4. Put items into the hopper.
5. Check inventory of dropper.

### Plugin list:
- Self-made plugin to set Dropper's MaxStackSize.

### Paper build number:
git-Paper-221 (MC: 1.14.4)

### Anything else:
- The problem also occurs when transferring items using droppers into the container.
- I have set `disable-move-event` to true for hoppers.

### LINKED ISSUE #6329: Composter dont work like expected with bigger hopper
### Expected behavior

If for example 5 items get transported by an hopper that the composter manages them correctly.

### Observed/Actual behavior

If 5 items get pushed into a composter, it go only by 1 row higher, manually with hand it is 5 times faster so I believe the composter registers only 1 and not 5 items.

### Steps/models to reproduce

Hopper connected with a composter
Change paper config setting from 1 item per push to a higher level (the higher the clearer)
Look at the composter 

### Plugin list

No plugin

### Paper version

This server is running Paper version git-Paper-156 (MC: 1.17.1) (Implementing API version 1.17.1-R0.1-SNAPSHOT) (Git: d239eca)
You are running the latest version
Previous version: git-Paper-92 (MC: 1.17.1)

### Agreements

- [X] I am running the latest version of Paper available from https://papermc.io/downloads.
- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] My version of Minecraft is supported by Paper.

### Other

_No response_

### PR DESCRIPTION
**Pull request by Machine-Maker**
closes #2662 

I know the hopper code is prickly cause you want it to be as performant as possible. Don't think this change really adds much if anything. 

I tested it both pushing into other inventories, and pulling from other inventories.

Fixes #6329

### CODE REVIEW

**Comment by Proximyst** [comment]
Does this also work fine with dropping items into it?

**Comment by Machine-Maker** [comment]
@Proximyst I didn't test it before, but I just did and it does respect the stack size when dropping items in. I was testing with a max stack size of 1.

**Review by Proximyst — DISMISSED** [review]
Then it LGTM :)

**Comment by Phoenix616** [comment]
Does this keep the rest around correctly?

**Comment by Machine-Maker** [comment]
@Phoenix616 Can you rephrase that? I'm not sure I follow what you mean.

**Comment by Phoenix616** [comment]
The items that are not going to get transferred due to it reaching the max, do they stay around or do they get lost.

**Comment by Machine-Maker** [comment]
Oh yeah, they stick around. That subtract call in the conditional "removes" the item from the inventory.

**Comment by Phoenix616** [comment]
I see, good.

**Comment by Phoenix616** [comment]
Another thing I just though about: If you increase the hopper transfer amount the current code will (as far as I understand it) not pull any items at all even though it would be possible   to pull a partial amount.

Something like checking how much will fit and only moving that amount would be nice. 

Edit: It looks like the best overall solution here would be to do `int j = Math.min(itemstack.getMaxStackSize(), iinventory1.getMaxStackSize()) - itemstack1.getCount();` on [line 12](https://github.com/PaperMC/Paper/blob/c9a710a9a7aea2073514635fa12c445dca6c6b42/Spigot-Server-Patches/0612-Make-hoppers-respect-inventory-max-stack-size.patch#L12) instead of the check afterwards.

**Comment by Machine-Maker** [comment]
Good catch, yeah, right now if I set hopper amount to 5 in spigot.yml, it pushes in 5, but only puts 1 in each slot in the inventory below (that has max stack size of 1)

**Comment by Machine-Maker** [comment]
So yeah, phoenix, your fix was *part* of the solution, but I had to add a bit more above it relating to the code when its moving into an empty slot. Should work all around now.

**Comment by Phoenix616** [comment]
While this probably works fine I'm really not a fan of always cloning the full item stack. That's some expensive operation that should only really be done when necessary (which in this case should only be when the moved count wasn't the same as originally? Not sure why a clone is needed there if the vanilla code should already be able to handle partially transfered amounts due to max stack size?)

**Comment by Machine-Maker** [comment]
Yeah, so the clone is necessary, because vanilla mechanics means there is **never** going to be a ItemStack going into an empty slot that **isn't** going to be accepted. Vanilla mechanics only move 1 item at a time, and all empty slots are able to accept at least 1 item, if only 1 item. Here, we run into situations where you are moving 5 items at once (an ItemStack of count 5) into an empty slot that only holds 2 (set by the maxstacksize). Vanilla mechs would move that itemstack, then set the original itemstack to the NULL_ITEM. we can't do that here because we need to return the leftover itemstack (in this case 3)

**Comment by Machine-Maker** [comment]
Although, looking at it again, my original concern might've been invalid in the first place. For some reason, I thought that the line that sets the itemstack = ItemStack.b (NULL_ITEM) would also set the itemstack that was moved in the previous operation to null, idk why I thought that since its not an operation on that itemstack object. I'll go back and test w/o the clone

**Comment by Machine-Maker** [comment]
Yeah, works totally find w/o the clone, not sure why I thought I would need it :shrug:

**Comment by Machine-Maker** [comment]
Still think this is valid

**Comment by Machine-Maker** [comment]
`Rebased after the NMS repackage`

**Comment by Machine-Maker** [comment]
Rebased for 1.17.1

**Review by Proximyst — APPROVED** [review]
